core/drivers/crypto/crypto_api/include/drvcrypt.h

```diff
@@ -0,0 +1,51 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2018-2019 NXP
+ *
+ * Brief   Crypto Driver exported constants and interfaces.
+ */
+#ifndef __DRVCRYPT_H__
+#define __DRVCRYPT_H__
+
+#include <tee_api_types.h>
+#include <trace.h>
+
+#ifdef CFG_CRYPTO_DRIVER_DEBUG
+#define CRYPTO_TRACE	DMSG
+#else
+#define CRYPTO_TRACE(...)
+#endif
+
+/*
+ * Crypto Library Algorithm enumeration
+ */
+enum drvcrypt_algo_id {
+	CRYPTO_HASH = 0,     /* HASH driver */
+	CRYPTO_MAX_ALGO      /* Maximum number of algo supported */
+};
+
+/*
+ * Registers the Cryptographic's operation in the table of modules
+ *
+ * @algo_id  ID of the Cryptographic module
+ * @ops      Operation (function/structure) to register
```

Comment:
I feel this operation table reference is not clear.
In the case of a hash driver, according to this P-R, one shall register a context allocation function.
For other crypto algorithm, could it be another function, or a table of function pointers?

How to ensure one registers a compliant reference?

Maybe drvcrpyt_hash driver should provide an API function `drvcrypt_hash_register()` that expects an explicitly typed `ops` (I.e. `hw_hash_allocate` for hash drivers) and that would call generic `drvcrypt_register()` with reference made `void *` opaque.

Reply:
This table will be increased when the other HW crypto operations will be included in the source. This table should be kept as `void *` because algorithms will not registered the same ops. This is under the responsibility of the developer to ensure that ops register is correct, if not it means that no tests were run.
Goal of this table is to plug any Crypto HW driver (not only NXP CAAM driver). The objective is to make low level HW driver interface common in that way the crypto_api doesn't need to be change and include the common logic not needed to be duplicated in each low level HW driver.
Many options exist to do this. I chose to let the low level HW driver to register itself and as it's platform specific and it's able to decide depending on HW capabilities if it can be use or not.

Reply:
Suggestion:  add to `drvcrypt_hash.h`
```C
static inline TEE_Result drvcrypt_register_hash(hw_hash_allocate allocate)
{
	return drvcrypt_register(DRVCRYP_HASH, (void *)allocate);
}
```

hw_hash_allocate

Reply:
ok